core/store/src/trie/mem/resharding.rs

```diff
@@ -0,0 +1,250 @@
+use crate::{NibbleSlice, TrieChanges};
+
+use super::arena::ArenaMemory;
+use super::updating::{MemTrieUpdate, OldOrUpdatedNodeId, TrieAccesses, UpdatedMemTrieNode};
+use itertools::Itertools;
+use near_primitives::types::AccountId;
+use std::ops::Range;
+
+/// Whether to retain left or right part of trie after shard split.
+pub enum RetainMode {
+    Left,
+    Right,
+}
+
+/// Decision on the subtree exploration.
+#[derive(Debug)]
+enum RetainDecision {
+    /// Retain the whole subtree.
+    RetainAll,
+    /// The whole subtree is not retained.
+    DiscardAll,
+    /// Descend into all child subtrees.
+    Descend,
+}
+
+impl<'a, M: ArenaMemory> MemTrieUpdate<'a, M> {
+    /// Splits the trie, separating entries by the boundary account.
+    /// Leaves the left or right part of the trie, depending on the retain mode.
+    ///
+    /// Returns the changes to be applied to in-memory trie and the proof of
+    /// the split operation. Doesn't modifies trie itself, it's a caller's
+    /// responsibility to apply the changes.
+    pub fn retain_split_shard(
+        self,
+        _boundary_account: AccountId,
+        _retain_mode: RetainMode,
+    ) -> (TrieChanges, TrieAccesses) {
+        // TODO(#12074): generate intervals in nibbles.
+
+        self.retain_multi_range(&[])
+    }
+
+    /// Retains keys belonging to any of the ranges given in `intervals` from
+    /// the trie.
+    ///
+    /// Returns changes to be applied to in-memory trie and proof of the
+    /// retain operation.
+    fn retain_multi_range(mut self, intervals: &[Range<Vec<u8>>]) -> (TrieChanges, TrieAccesses) {
+        debug_assert!(intervals.iter().all(|range| range.start < range.end));
+        let intervals_nibbles = intervals
+            .iter()
+            .map(|range| {
+                NibbleSlice::new(&range.start).iter().collect_vec()
+                    ..NibbleSlice::new(&range.end).iter().collect_vec()
+            })
+            .collect_vec();
+        // let root = self.get_root().unwrap();
+        // TODO(#12074): consider handling the case when no changes are made.
+        // TODO(#12074): restore proof as well.
+        self.retain_multi_range_recursive(0, vec![], &intervals_nibbles);
+        self.to_trie_changes()
+    }
+
+    /// Recursive implementation of the algorithm of retaining keys belonging to
+    /// any of the ranges given in `intervals` from the trie.
+    ///
+    /// `node_id` is the root of subtree being explored.
+    /// `key_nibbles` is the key corresponding to `root`.
+    /// `intervals_nibbles` is the list of ranges to be retained.
+    ///
+    /// Returns id of the node after retain applied.
```

Comment:
nit: comment no longer hold, can maybe mention, all changes are tracked in `self.updated_nodes`?